spec/ics-2-consensus-requirements/README.md

```diff
@@ -0,0 +1,76 @@
+If a blockchain has deterministic safety(as opposed to probabilistic safety in Nakamoto consensus), then there cannot be more than one child for each block. In Tendermint, this assumption breaks when +1/3 validators are malicious, producing multiple blocks those are direct children of a single block and all of them are verifiable with the parent. It makes conflicting packets delivered from the failed chain, so for the chains who are receiving from this chain, **fraud proof #10** mechanism should be applied in this case. However, he failed chain itself also need to be recovered and reconnected again. It will be covered in **byzantine recovery strategies #6**.
+
+2. Headers have at least one direct child
+There is at least one direct child for all blocks, meaning that the lightclient logic can proceed the blocks one by one even in the worst case. If it not satisfied then there can be a point where the lightclient stops and cannot proceed, which halts IBC connection unexpectedly. This also can be violated when the blocks are restarted out of the consensus, for example in Tendermint, (1/2 < validators < 2/3) forked out the chain. This also will be covered in **byzantine recovery strategies**.
+
+3. If a block verifies another block then it is a descendant of the block
+Lightclient should not verify packets which are not in its chain. If the verifier returns true for a block that is not a descendant, it simply means that there is an error in the lightclient logic.
+
+4. Header can have a state only if it is a valid transition from the parent's(see the next paragraph)
+Header cannot have a state which is not a transition from its parent. This requires that the chain satisfies the IBC protocol and runs it honestly. In case where the validators become malicious and send invalid messages, the application logic on the destination chain should prevent further interchain infection.
+ 
+These requirements allow channels to work safely without concerning about double spending attack. 
+If a block is submitted to the chain(and optionally passed some challenge period for fraud proof), then it is assured that the packet is finalized so the application logic can process it.
+
+### Technical Specification
+
+In a real world, `Header`s do not have to contain the full state data; it can contain an aggregated data set identifier, for example a merkle root, where the actual data latter given can be proven that it is a member of that set. If we don't consider about the data availability problem, this partial blocks, which is be called `Header`s, still satisfies `Header`.
+
+Following functions exist over `Header`s.
+
+* `height : Header -> int`
+Returns the height of the block.
+
+* `verify : Header -> Header -> Maybe<LightClientProof> -> Maybe<error>`
+Verifies the latter block is a descendant of the former block, as defined in **lightclient specification**. Can take an additional lightclient proof argument.
+
+* `state : Header -> [MerkleProof] -> Error | KVStore`
```

Comment:
Took a bit to see how we get a KVStore, but I see this as a "partial" KV Store representing a proveable subset of the entire blockchain state the Header refers to. With that understanding, I like it quite a bit. Very elegant representation

Reply:
We should disambiguate this with the previous definition of a state as `s: []byte->[]byte`

Reply:
I don't understand how we can return a `KVStore` here, since we don't actually have the data, it's returned on-demand.

Additionally, I think we should abstract `MerkleProof` to `Accumulator`, but keep the key-value abstraction.

Maybe `verifySubstate : RootOfTrust -> Key -> Value -> Maybe Error`

We might need slightly different semantics for range exclusion proofs, which I think we should include in these requirements. Ideally the function can encapsulate both.

Reply:
(where `type Key = type Value = []byte`)

Reply:
If we assume MerkleProofs come with Key and Value, this is much nicer for the data availability problem.

The point I assume is that we get a trusted header and N `MerkleProofs`, each containing `{Key, Value, HashPathToRoot}`. We can verify them and, on succcess, return a partial KVStore, containing only the keys related to the proofs provided. This is guaranteed to be a subset of the complete KVStore on the blockchain that match the header.

Reply:
From the discussion: what we are using from the header is essentially the accumulator root. Headers can implement function `getRoot`, which simply returns the accumulator root, and the rest can be moved into the other ICS for accumulators. This will make the interface as minimal as possible.